### ISSUE TITLE
LE support for emulating memory operations in lib/sstep.c

### ISSUE THREAD

**Issue by ajdlinux** [linuxppc/linux#22]
Currently, the single step emulation library in arch/powerpc/lib/sstep.c doesn't implement emulation for certain classes of memory operations on little-endian systems - there's an awful lot of code along the lines of:

```
case LOAD_MULTI:
    if (regs->msr & MSR_LE)
        return 0;
```

We should make this code work on LE as it's rather useful for various forms of debugging.

**Comment by ajdlinux** [comment]
I'm going to try and get some work done on this. Would be nice to have a test suite of some description, not entirely sure what the best way to do that would be.

**Comment by ajdlinux** [comment]
From conversations with various people at OzLabs - it appears that a lot of LE support code was added as part of a patchset in 2013 (https://lists.ozlabs.org/pipermail/linuxppc-dev/2013-November/112925.html). However, the patch removing the endianness check in emulate_step() wasn't merged for reasons that are unclear.

Hopefully, we should be able to just remove the endianness checks and everything will just work, but we probably want to test that the results are the same. Suggestions welcome on how to write a test suite.

**Comment by mpe** [comment]
As far as testing, `emulate_step()` is used by the kprobes handler. So you could probably hand write some assembler functions that exercise the instructions you're interested in, and then have a kernel module that places kprobes on those instructions.

The test would need to be on two levels, firstly that you don't oops, and secondly that the emulation of the instruction actually did what was required.

**Comment by mpe** [comment]
Fixed in e148bd17f48b ("powerpc: Emulation support for load/store instructions on LE"), and d955189ae427 ("powerpc: Handle opposite-endian processes in emulation code").
